polaris-migrator/src/utilities/sass.ts

```diff
@@ -251,3 +263,387 @@ export function createInlineComment(text: string, options?: {prose?: boolean}) {
+          // Normally, this comes from stylelint config, but for this shim we
+          // just hard-code it, and instead rely on the "seconary" options
+          // object for passing through the jscodeshift options.
+          true,
+          {
+            ...options,
+            partialFixStrategy: 'report',
+            __injectReportsAsComments: true,
+          },
+          {fix: true},
+        );
+
+        ruleProcessor(root);
+      },
+    };
+
+    return postcss(plugin).process(fileInfo.source, {
+      syntax: require('postcss-scss'),
+    }).css;
+  };
+}
+
+export function createStylelintRule(ruleName: string, ruleFn: PolarisMigrator) {
+  const wrappedRule: StylelintRule = ((
+    primary,
+    {
+      partialFixStrategy = 'fix',
+      __injectReportsAsComments = false,
+      ...secondaryOptions
+    }: PluginOptions,
+    context,
+  ) => {
+    const reports = new Map<PostCSSNode, Report[]>();
+    const suggestions = new Map<PostCSSNode, string>();
+
+    const addDedupedReport = (newReport: Report) => {
+      if (!reports.has(newReport.node)) {
+        reports.set(newReport.node, []);
+      }
+
+      const reportsForNode = reports.get(newReport.node)!;
+
+      if (
+        reportsForNode.findIndex(
+          (existingReport) =>
+            existingReport.severity === newReport.severity &&
+            existingReport.message === newReport.message,
+        ) === -1
+      ) {
+        reportsForNode.push(newReport);
+      }
+    };
+
+    const flushReportsAsComments = () => {
+      // @ts-expect-error No idea why TS is complaining here
+      for (const [node, reportsForNode] of reports) {
+        node.before(
+          createInlineComment(POLARIS_MIGRATOR_COMMENT, {prose: true}),
+        );
+
+        for (const report of reportsForNode) {
+          node.before(
+            createInlineComment(`${report.severity}: ${report.message}`, {
+              prose: true,
+            }),
+          );
+        }
+      }
+      reports.clear();
+
+      // @ts-expect-error No idea why TS is complaining here
+      for (const [node, suggestion] of suggestions) {
+        node.before(createInlineComment(suggestion));
+      }
+      suggestions.clear();
+    };
+
+    // TODO: When moving to style-lint, migrate this to call style-lint's
+    // stylelint.utils.report()
+    const flushReportsToStylelint = flushReportsAsComments;
+
+    function createWalker<T extends PostCSSNode>(args: {
+      walker: (node: T) => false | void;
+      mutableKeys?: (keyof T)[];
+      serialiseSuggestion: (node: T) => string;
+    }): (node: T) => false | void;
+    function createWalker<T extends PostCSSNode>(args: {
+      valueParserKey: keyof T;
+      walker: (node: T, parsedValue: ParsedValue) => false | void;
+      mutableKeys?: (keyof T)[];
+      serialiseSuggestion: (node: T) => string;
+    }): (node: T) => false | void;
+    function createWalker<T extends PostCSSNode>(args: {
```

Comment:
This factory function wraps up common logic across all of our migrators:

- Tracking if changes have occurred to a postcss Node (previously done manually in each migration via a `targets` array)
- Adding comments if there's a partial migration
- Add comments when there's an error/warning detected, but it's unmigratable